infobind/cache.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/sqlexec"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	using   = "using"
+	deleted = "deleted"
+)
+
+// bindRecord store the basic bind info and bindSql astNode.
+type bindRecord struct {
+	*bindMeta
+	ast ast.StmtNode //the ast will be used to do query sql bind check
+}
+
+// bindCache holds a bindDataMap
+//   key: original sql
+//   value: a slice of bindRecords.
+type bindCache map[string][]*bindRecord
+
+// Handler hold an atomic bindCache.
+type Handler struct {
+	atomic.Value
+}
+
+// BindCacheUpdater is used to update the global bindCache.
+// BindCacheUpdater will update the bind cache pre 3 second in domain gorountine loop. When the tidb server first startup, the updater will load all bind info in memory; then load diff bind info pre 3 second.
+type BindCacheUpdater struct {
+	ctx sessionctx.Context
+
+	parser         *parser.Parser
+	lastUpdateTime types.Time
+	globalHandler  *Handler
+}
+
+type bindMeta struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	// Status will only be 0 or 1:
+	//   deleted: bindMeta has been marked as deleted status.
+	//   using: bindMeta is in using status.
+	Status     string
+	CreateTime types.Time
+	UpdateTime types.Time
+	Charset    string
+	Collation  string
+}
+
+// NewBindCacheUpdater create a new BindCacheUpdater.
+func NewBindCacheUpdater(ctx sessionctx.Context, handler *Handler, parser *parser.Parser) *BindCacheUpdater {
+	return &BindCacheUpdater{
+		globalHandler: handler,
+		parser:        parser,
+		ctx:           ctx,
+	}
+}
+
+// NewHandler create a Handler with a bindCache.
+func NewHandler() *Handler {
+	handler := &Handler{}
+	return handler
+}
+
+// Get get bindCache from a Handler.
+func (h *Handler) Get() bindCache {
+	bc := h.Load()
+
+	if bc != nil {
+		return bc.(map[string][]*bindRecord)
+	}
+
+	return make(map[string][]*bindRecord)
+}
+
+// LoadDiff use to load new bind info to bindCache bc.
+func (bindCacheUpdater *BindCacheUpdater) loadDiff(sql string, bc bindCache) error {
+	recordSets, err := bindCacheUpdater.ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := recordSets[0]
+	defer terror.Call(rs.Close)
+	chkBatch := rs.NewRecordBatch()
+	for {
+		err = rs.Next(context.TODO(), chkBatch)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chkBatch.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chkBatch.Chunk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record := newBindMeta(row)
+			err = bc.appendNode(record, bindCacheUpdater.parser)
+			if err != nil {
+				return err
+			}
+
+			if record.UpdateTime.Compare(bindCacheUpdater.lastUpdateTime) == 1 {
+				bindCacheUpdater.lastUpdateTime = record.UpdateTime
+			}
+		}
+	}
+}
+
+// Update update the BindCacheUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false.
```

Comment:
Update updates the BindCacheUpdater's bindCache. The `fullLoad` is true when tidb first startup, otherwise it is false.